### PR TITLE
Feature/rebuild ir unit test

### PR DESCRIPTION
**Pull request by Hejunchao6**

### CODE REVIEW

File: src/Nncase.Tests/Core/UnitTestDumpUtility.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    [Fact]
+    public void TestDumpUtility()
+    {
+        DumpUtility.WriteResult("./test3", "1");
+        DumpUtility.WriteResult<int>("./test3", new[] { 1 });
+        DumpUtility.SerializeShape(new[] { 1, 1, 1 });
+        DumpUtility.PathJoinByCreate("./", "test4");
+        DumpUtility.WriteBinFile("./test5", new Tensor<int>(new[] { 1 }));
+    }
+
```
Comment (Hejunchao6) [inline src/Nncase.Tests/Core/UnitTestDumpUtility.cs:32]: 1

File: src/Nncase.Tests/Core/UnitTestGetReplaceUtility.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            stride: new[] { 1, 1 },
+            padding: Tensor.From<int>(new int[] { 1, 1, 1, 1 }, new[] { 2, 2 }),
+            dilation: new[] { 1, 1 },
+            PadMode.Constant,
+            1);
+        var a = Utility.WithTmpBF16(_ => expr);
+        var b = Utility.WithTmpType(_ => expr, DataTypes.Float32);
+        var c = Utility.WithTmp4DShape(_ => expr, new[] { 1, 1, 1, 1 });
+        var d = Utility.Get4DGNNEShape(new[] { 0, 1, 2, 3 });
+        Assert.Throws<InvalidOperationException>(() => Utility.Get4DGNNEShape(new[] { 0, 1, 2, 3, 4 }));
```
Comment (Hejunchao6) [inline src/Nncase.Tests/Core/UnitTestGetReplaceUtility.cs:37]: 2

File: src/Nncase.Tests/Core/UnitTestPatternPrinter.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+namespace Nncase.Tests.CoreTest;
+
+public sealed class UnitTestPatternPrinter
+{
+    [Fact]
+    public void TestDumpAsIL()
+    {
+        IPattern pattern = null!;
+        Assert.Throws<NullReferenceException>(() => PatternPrinter.DumpAsIL(pattern));
+    }
```
Comment (Hejunchao6) [inline src/Nncase.Tests/Core/UnitTestPatternPrinter.cs:21]: 3

File: src/Nncase.Tests/Evaluator/UnitTestCostEvaluate.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        costEvaluateContext.CurrentCall = loadCall;
+        Assert.Equal(loadCall, costEvaluateContext.CurrentCall);
+    }
+
+    [Fact]
+    public void TestCostEvaluateVisitor()
+    {
+        var costEvaluateVisitor1 = new CostEvaluateVisitor();
+        var costEvaluateVisitor2 = new CostEvaluateVisitor();
+        Assert.False(costEvaluateVisitor1.Equals(costEvaluateVisitor2));
```
Comment (Hejunchao6) [inline src/Nncase.Tests/Evaluator/UnitTestCostEvaluate.cs:46]: 4

File: src/Nncase.Tests/Rules/Neutral/UnitTestSimplifyFactory.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+namespace Nncase.Tests.Rules.NeutralTest;
+
+public class UnitTestSimplifyFactory
+{
+    [Fact]
+    public void TestSimplifyFactory()
+    {
+        Assert.NotEqual(new(), SimplifyFactory.SimplifyAdd());
+        Assert.NotEqual(new(), SimplifyFactory.SimplifyMul());
+    }
```
Comment (Hejunchao6) [inline src/Nncase.Tests/Rules/Neutral/UnitTestSimplifyFactory.cs:22]: 5

File: src/Nncase.Tests/Core/UnitTestBufferAllocation.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+{
+    [Fact]
+    public void TestBufferAllocation()
+    {
+        var bufferAllocation = new BufferAllocation(default, DataTypes.Int8, 1UL, 1UL, 1UL, new[] { 1, 1, 1, 1 }, new[] { 1, 1, 1, 1 }, new[] { 1, 1, 1, 1 });
+        Assert.Equal(default, bufferAllocation.MemoryLocate);
+        Assert.Equal(1UL, bufferAllocation.SharedModule);
+        Assert.Equal(1UL, bufferAllocation.Start);
+        Assert.Equal(DataTypes.Int8, bufferAllocation.DType);
+        Assert.Equal(1UL, bufferAllocation.Size);
```
Comment (xhuohai) [inline src/Nncase.Tests/Core/UnitTestBufferAllocation.cs:27]: {1,1,1,1} is too simple

File: src/Nncase.Tests/Core/UnitTestDumpUtility.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        Assert.True(Directory.Exists("./test4"));
+        Assert.True(File.Exists("./test5"));
+    }
+
+    [Fact]
+    public void TestBinFileUtil()
+    {
+        BinFileUtil.WriteBinInputs(new Tensor[] { new Tensor<int>(new[] { 1 }) }, "./");
+        BinFileUtil.WriteBinOutputs(new Tensor[] { new Tensor<int>(new[] { 1 }) }, "./");
+        Assert.True(Directory.Exists("./"));
```
Comment (xhuohai) [inline src/Nncase.Tests/Core/UnitTestDumpUtility.cs:44]: ./ is always true,  should use another dir

File: src/Nncase.Tests/Core/UnitTestTIR.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace Nncase.Tests.CoreTest;
+
+public sealed class UnitTestTIR
+{
+    [Fact]
+    public void TestLogicalBuffer()
+    {
+        var logicalBuffer = new LogicalBuffer("logicalBuffer", default, new TensorConst(new Tensor<int>(new[] { 1 })));
+        Assert.Equal(logicalBuffer.Length, logicalBuffer.Length);
```
Comment (xhuohai) [inline src/Nncase.Tests/Core/UnitTestTIR.cs:24]: this line is useless

File: src/Nncase.Tests/Core/UnitTestWeakList.cs [hunk 9]
```diff
@@ -26,5 +27,64 @@ public void TestWeakList()
         weakList.Add(3);
         weakList.Add(4);
         Assert.Equal(new WeakReference<object>(1).TryGetTarget(out _), weakList.GetWeakReference(0).TryGetTarget(out _));
+
+#if false
```
Comment (xhuohai) [inline src/Nncase.Tests/Core/UnitTestWeakList.cs:31]: why leave it with false?

File: src/Nncase.Tests/Core/UnitTestWeakList.cs [hunk 10]
```diff
@@ -26,5 +27,64 @@ public void TestWeakList()
...
+
+        var sizeBefore = list.ToList().Count;
+
+        // Act
+        _ = list.GetEnumerator();
+        var sizeAfter = list.ToList().Count;
+
+        // Assert
+        Assert.Equal(sizeAfter, sizeBefore);
+        Assert.True(sizeAfter > sizeBefore / 4);
```
Comment (xhuohai) [inline src/Nncase.Tests/Core/UnitTestWeakList.cs:88]: always true if line 87 is true

File: src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs [hunk 11]
```diff
@@ -365,6 +365,26 @@ public void TestSlice()
...
+        Assert.True(expr.InferenceType());
+        Assert.Equal(expect, expr);
+    }
+
+    [Fact]
+    public void TestNHWCToWNCH()
+    {
+        var input = Tensor.From<int>(Enumerable.Range(0, 120).ToArray(), new Shape(new[] { 2, 3, 4, 5 }));
+        var expr = NHWCToWNCH(input);
+        var expect = Transpose(input, new[] { 3, 0, 1, 2 });
```
Comment (xhuohai) [inline src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs:386]: NHWCToWNCH is {2 0 3 1}

File: src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs [hunk 12]
```diff
@@ -658,6 +678,62 @@ public void TestWhere()
         Assert.Equal(expectTF, exprTF.Evaluate().AsTensor());
     }
 
+    [Fact]
+    public void TestReduceMean()
+    {
+        long axis = 0L;
+        long[] keepDims = { 0L, 1L };
```
Comment (xhuohai) [inline src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs:685]: keepdims is a interger

File: src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs [hunk 13]
```diff
@@ -658,6 +678,62 @@ public void TestWhere()
...
+        var expect = Reduce(ReduceOp.Mean, expr_a, axis, 0f, keepDims);
+        CompilerServices.InferenceType(expect);
+        Assert.Equal(expr, expect);
+    }
+
+    [Fact]
+    public void TestReduceMin()
+    {
+        long axis = 0L;
+        long[] keepDims = { 0L, 1L };
```
Comment (xhuohai) [inline src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs:699]: same as above

File: src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs [hunk 14]
```diff
@@ -658,6 +678,62 @@ public void TestWhere()
...
+        var expect = Reduce(ReduceOp.Max, expr_a, axis, 0f, keepDims);
+        CompilerServices.InferenceType(expect);
+        Assert.Equal(expr, expect);
+    }
+
+    [Fact]
+    public void TestReduceSum()
+    {
+        long axis = 0L;
+        long[] keepDims = { 0L, 1L };
```
Comment (xhuohai) [inline src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs:727]: same as above

File: src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs [hunk 15]
```diff
@@ -658,6 +678,62 @@ public void TestWhere()
...
+        var expect = Reduce(ReduceOp.Min, expr_a, axis, 0f, keepDims);
+        CompilerServices.InferenceType(expect);
+        Assert.Equal(expr, expect);
+    }
+
+    [Fact]
+    public void TestReduceMax()
+    {
+        long axis = 0L;
+        long[] keepDims = { 0L, 1L };
```
Comment (xhuohai) [inline src/Nncase.Tests/Evaluator/UnitTestEvaluatorTensors.cs:713]: same as above

File: src/Nncase.Tests/Rules/Neutral/UnitTestSimplifyFactory.cs [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+    private static readonly List<IRewriteRule> _simplifyMul = new()
+    {
+        Rewrite(X * 1, X),
+        Rewrite(1 * X, X),
+    };
+
+    [Fact]
+    public void TestSimplifyFactory()
+    {
+        Assert.NotEqual(_simplifyAdd, SimplifyFactory.SimplifyAdd());
```
Comment (xhuohai) [inline src/Nncase.Tests/Rules/Neutral/UnitTestSimplifyFactory.cs:41]: why use not equal?

File: src/Nncase.Tests/Core/UnitTestWeakList.cs (hunk 9 above)
Comment (Hejunchao6) [reply src/Nncase.Tests/Core/UnitTestWeakList.cs:31]: The class is't the real weaklist,"GC.(null)" dose‘t work.No sure if it will be fixed after.

**Review by xhuohai — APPROVED** [review]

**Pull request merged**